samples/net/openthread/coap_client/src/coap_utils.c

```diff
@@ -0,0 +1,267 @@
+		}
+	}
+
+	if (payload) {
+		ret = coap_packet_append_payload_marker(request);
+		if (ret < 0) {
+			LOG_ERR("Unable to append payload marker");
+			goto end;
+		}
+
+		ret = coap_packet_append_payload(request, payload, payloadSize);
+		if (ret < 0) {
+			LOG_ERR("Not able to append payload");
+			goto end;
+		}
+	}
+
+end:
+	return ret;
+}
+
+static int coap_send_message(const char *dstAddr, struct coap_packet *request)
+{
+	struct sockaddr_in6 addr6;
+
+	addr6.sin6_family = AF_INET6;
+	addr6.sin6_port = htons(PEER_PORT);
+	addr6.sin6_scope_id = 0U;
+
+	inet_pton(AF_INET6, dstAddr, &addr6.sin6_addr);
+
+	return sendto(sock, request->data, request->offset, 0,
+		      (struct sockaddr *)&addr6, sizeof(addr6));
+}
+
+static int coap_set_response_callback(struct coap_packet *request,
+				      coap_reply_t replyCb)
+{
+	struct coap_reply *reply;
+	int ret = 0;
+
+	reply = coap_reply_next_unused(m_replies, COAP_MAX_REPLIES);
+	if (!reply) {
+		LOG_ERR("No resources for waiting for replies.");
+		ret = -ENOMEM;
+		goto end;
+	}
+
+	coap_reply_clear(reply);
+	coap_reply_init(reply, request);
+	reply->reply = replyCb;
+
+end:
+	return ret;
+}
+
+int coap_open_connection(void)
+{
+	sock = socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		LOG_ERR("Failed to create UDP socket %d", errno);
+		return -errno;
+	}
+
+	fds.fd = sock;
+	fds.events = POLLIN;
+
+	/* start sock receive thread */
+	k_thread_create(&receive_thread_data, receive_stack_area,
+			K_THREAD_STACK_SIZEOF(receive_stack_area),
+			(k_thread_entry_t)coap_receive, NULL, NULL, NULL,
+			/* Lowest priority cooperative thread */
+			K_PRIO_COOP(CONFIG_NUM_COOP_PRIORITIES - 1), 0,
+			K_NO_WAIT);
+	k_thread_name_set(&receive_thread_data, "CoAP-sock-recv");
+	LOG_DBG("CoAP socket receive thread started");
+
+	return 0;
+}
+
+void coap_close_connection(void)
+{
+	(void)close(sock);
+}
+
+int coap_send_non_con_request(enum coap_method method, const char *dstAddr,
+			      const char *const *uriPathOption, u8_t *payload,
+			      u16_t payloadSize)
+{
+	int ret;
+	struct coap_packet request;
+	u8_t buf[MAX_COAP_MSG_LEN];
+
+	ret = coap_init_request(method, uriPathOption, payload, payloadSize,
+				&request, buf);
+	if (ret < 0) {
+		goto end;
+	}
+
+	ret = coap_send_message(dstAddr, &request);
+	if (ret < 0) {
+		LOG_ERR("Transmission failed: %d", errno);
+		goto end;
+	}
+
+end:
+	return ret;
+}
+
+int coap_send_con_request(enum coap_method method, const char *dstAddr,
+			  const char *const *uriPathOption, u8_t *payload,
+			  u16_t payloadSize, coap_reply_t replyCb)
+{
+	int ret;
+	struct coap_packet request;
+	u8_t buf[MAX_COAP_MSG_LEN];
+
+	ret = coap_init_request(method, uriPathOption, payload, payloadSize,
+				&request, buf);
+	if (ret < 0) {
+		goto end;
+	}
+
+	ret = coap_set_response_callback(&request, replyCb);
+	if (ret < 0) {
+		goto end;
+	}
+
+	ret = coap_send_message(dstAddr, &request);
+	if (ret < 0) {
+		LOG_ERR("Transmission failed: %d", errno);
+		goto end;
```

Comment:
not needed goto

Reply:
I prefer common approach for error handling to prevent bugs

Reply:
I hate goto at all, but I seem to be the only one here:)
Leave it as it is.

Reply:
Just a qustion.. what is Zephyr overall policy on this? Are goto commonly used (e.g. via some macros or like that)?

Reply:
`goto` is allowed (and recommended) for error handling.

Reply:
In general Zephyr follows Linux kernel coding style. In the matter of 'goto' it says:

> The goto statement comes in handy when a function exits from multiple locations and some common work such as cleanup has to be done. If there is no cleanup needed then just return directly.
I haven't it known about it so far, I used 'goto' to provide that every function has at most one single point of exit at the end.  I know that e.g. MISRA has that. I will change it.

Reply:
I think the conclusion is to **use** goto for error handling. Are you using it in different scenarios?

Reply:
Nope, only error handlings